File: Elixir/2024/2024-004/2024-004.exs
defmodule WordSearch do
  @directions [
    {0, 1},   # Right
    {1, 0},   # Down
    {1, 1},   # Diagonal-right-down
    {1, -1},  # Diagonal-left-down
    {0, -1},  # Left
    {-1, 0},  # Up
    {-1, -1}, # Diagonal-left-up
    {-1, 1}   # Diagonal-right-up
  ]

  @target_word "XMAS"

  def read_grid(file_path) do
    File.read!(file_path)
    |> String.split("\n", trim: true)
    |> Enum.map(&String.graphemes/1)
  end

  def check_word(grid, word, x, y, dx, dy) do
    Enum.all?(0..(String.length(word) - 1), fn i ->
      nx = x + i * dx
      ny = y + i * dy

      if nx < 0 or ny < 0 or nx >= length(grid) or ny >= length(Enum.at(grid, 0)) do
        false
      else
        Enum.at(grid, nx) |> Enum.at(ny) == String.at(word, i)
      end
    end)
  end

  def count_occurrences(grid, word) do
    for x <- 0..(length(grid) - 1),
        y <- 0..(length(Enum.at(grid, 0)) - 1),
        {dx, dy} <- @directions,
        reduce: 0 do
      count ->
        if check_word(grid, word, x, y, dx, dy), do: count + 1, else: count
    end
  end

  def count_xmas_patterns(grid) do
    for x <- 1..(length(grid) - 2),
        y <- 1..(length(Enum.at(grid, 0)) - 2),
        reduce: 0 do
      count ->
        center = Enum.at(grid, x) |> Enum.at(y)
        top_left = Enum.at(grid, x - 1) |> Enum.at(y - 1)
        top_right = Enum.at(grid, x - 1) |> Enum.at(y + 1)
        bottom_left = Enum.at(grid, x + 1) |> Enum.at(y - 1)
        bottom_right = Enum.at(grid, x + 1) |> Enum.at(y + 1)

        if center == "A" and
             ((top_left == "M" and top_right == "S" and bottom_left == "M" and bottom_right == "S") or
              (top_left == "S" and top_right == "M" and bottom_left == "S" and bottom_right == "M") or
              (top_left == "M" and top_right == "M" and bottom_left == "S" and bottom_right == "S") or
              (top_left == "S" and top_right == "S" and bottom_left == "M" and bottom_right == "M")) do
          count + 1
        else
          count
        end
    end
  end

  def main do
    file_path = "/uploads/input.txt"
    grid = read_grid(file_path)

    xmas_count = count_occurrences(grid, @target_word)
    IO.puts("Count of XMAS occurrences: #{xmas_count}")

    pattern_count = count_xmas_patterns(grid)
    IO.puts("Total XMAS patterns: #{pattern_count}")
  end
end

WordSearch.main()

File: other-programming-languages/Spidermonkey/2024/2024-004/2024-004.js
// Function to read the grid from the input file
function readGrid(filePath) {
    try {
        // SpiderMonkey doesn't have 'fs', so use an environment-provided method like `read()` if available
        const data = read(filePath); // `read` is specific to some SpiderMonkey shells
        return data.split('\n').filter(line => line.trim() !== '');
    } catch (err) {
        print("Error reading file: " + err); // Use `print` instead of `console.log` in SpiderMonkey
        return [];
    }
}

// Function to check if the word exists in a given direction
function checkWord(grid, x, y, dx, dy, word, rows, cols) {
    for (let i = 0; i < word.length; i++) {
        const nx = x + i * dx;
        const ny = y + i * dy;

        if (nx < 0 || ny < 0 || nx >= rows || ny >= cols || grid[nx][ny] !== word[i]) {
            return false;
        }
    }
    return true;
}

// Count occurrences of the word "XMAS" in all directions
function countXMAS(grid) {
    const targetWord = "XMAS";
    const directions = [
        [0, 1],   // Right
        [1, 0],   // Down
        [1, 1],   // Diagonal-right-down
        [1, -1],  // Diagonal-left-down
        [0, -1],  // Left
        [-1, 0],  // Up
        [-1, -1], // Diagonal-left-up
        [-1, 1]   // Diagonal-right-up
    ];

    const rows = grid.length;
    const cols = grid[0].length;
    let count = 0;

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            for (const [dx, dy] of directions) {
                if (checkWord(grid, r, c, dx, dy, targetWord, rows, cols)) {
                    count++;
                }
            }
        }
    }

    return count;
}

// Count all X-MAS patterns
function countAllXMASPatterns(grid) {
    const rows = grid.length;
    const cols = grid[0].length;
    let count = 0;

    for (let r = 1; r < rows - 1; r++) {
        for (let c = 1; c < cols - 1; c++) {
            const center = grid[r][c];
            const topLeft = grid[r - 1][c - 1];
            const topRight = grid[r - 1][c + 1];
            const bottomLeft = grid[r + 1][c - 1];
            const bottomRight = grid[r + 1][c + 1];

            if (center === 'A') {
                // Pattern 1: M.S
                if (topLeft === 'M' && topRight === 'S' && bottomLeft === 'M' && bottomRight === 'S') count++;
                // Pattern 2: S.M
                if (topLeft === 'S' && topRight === 'M' && bottomLeft === 'S' && bottomRight === 'M') count++;
                // Pattern 3: M.M
                if (topLeft === 'M' && topRight === 'M' && bottomLeft === 'S' && bottomRight === 'S') count++;
                // Pattern 4: S.S
                if (topLeft === 'S' && topRight === 'S' && bottomLeft === 'M' && bottomRight === 'M') count++;
            }
        }
    }

    return count;
}

// Main function
function main() {
    const filePath = '/uploads/input.txt';
    const grid = readGrid(filePath).map(line => line.trim());

    if (grid.length === 0) {
        print("Error: Grid is empty or invalid.");
        return;
    }

    const xmasCount = countXMAS(grid);
    print(`Count of XMAS: ${xmasCount}`);

    const xmasPatterns = countAllXMASPatterns(grid);
    print(`Total X-MAS patterns: ${xmasPatterns}`);
}

// Run the main function
main();
